File: python/lib/scanner.py
"""Folder walking and file collection."""

from __future__ import annotations

from pathlib import Path
from typing import Generator

# Folders and files excluded by default (see SPEC §5)
DEFAULT_EXCLUDED_DIRS: frozenset[str] = frozenset(
    {".git", ".hg", ".svn", "node_modules", "__pycache__", ".mypy_cache", ".tox", ".venv", "venv"}
)

DEFAULT_EXCLUDED_FILES: frozenset[str] = frozenset(
    {"Thumbs.db", "desktop.ini", ".DS_Store"}
)

DEFAULT_EXCLUDED_SUFFIXES: frozenset[str] = frozenset({".pyc", ".pyo"})


def iter_files(
    root: Path,
    *,
    recursive: bool = False,
    include_hidden: bool = False,
    extensions: list[str] | None = None,
    min_size: int = 0,
    max_size: int | None = None,
) -> Generator[Path, None, None]:
    """Yield file paths under `root` that match the given filters.

    Args:
        root: Directory to scan.
        recursive: If True, descend into subdirectories.
        include_hidden: If True, include hidden files and system folders.
        extensions: If provided, only yield files with these extensions (e.g. ['.txt', '.csv']).
            Extensions are matched case-insensitively.
        min_size: Skip files smaller than this many bytes.
        max_size: Skip files larger than this many bytes. None means no limit.

    Yields:
        Path objects for each matching file.
    """
    extensions_lower = {ext.lower() if ext.startswith(".") else f".{ext.lower()}" for ext in extensions} if extensions else None

    def _should_skip_dir(d: Path) -> bool:
        name = d.name
        if not include_hidden and name.startswith("."):
            return True
        if name in DEFAULT_EXCLUDED_DIRS:
            return True
        return False

    def _should_skip_file(f: Path) -> bool:
        name = f.name
        if not include_hidden and name.startswith("."):
            return True
        if name in DEFAULT_EXCLUDED_FILES:
            return True
        if f.suffix.lower() in DEFAULT_EXCLUDED_SUFFIXES:
            return True
        if extensions_lower and f.suffix.lower() not in extensions_lower:
            return True
        return False

    def _walk(directory: Path) -> Generator[Path, None, None]:
        try:
            entries = sorted(directory.iterdir(), key=lambda p: (p.is_dir(), p.name.lower()))
        except PermissionError:
            return

        for entry in entries:
            if entry.is_symlink():
                continue
            if entry.is_dir():
                if recursive and not _should_skip_dir(entry):
                    yield from _walk(entry)
            elif entry.is_file():
                if _should_skip_file(entry):
                    continue
                try:
                    size = entry.stat().st_size
                except OSError:
                    yield entry  # Let detector handle the error
                    continue
                if size < min_size:
                    continue
                if max_size is not None and size > max_size:
                    continue
                yield entry

    yield from _walk(root)

File: nodejs/lib/scanner.js
/**
 * @file scanner.js
 * Folder walking and file collection using Node.js built-in fs/promises.
 * No third-party dependencies.
 */

import { readdir, stat } from "fs/promises";
import { join, resolve } from "path";

/** Directories excluded from scanning by default (see SPEC §5). */
const EXCLUDED_DIRS = new Set([
  ".git",
  ".hg",
  ".svn",
  "node_modules",
  "__pycache__",
  ".mypy_cache",
  ".tox",
  ".venv",
  "venv",
]);

/** Filenames excluded by default. */
const EXCLUDED_FILES = new Set(["Thumbs.db", "desktop.ini", ".DS_Store"]);

/** File extensions excluded by default. */
const EXCLUDED_EXTS = new Set([".pyc", ".pyo"]);

/**
 * @typedef {Object} ScanOptions
 * @property {boolean}        [recursive=false]      - Scan subdirectories.
 * @property {boolean}        [includeHidden=false]  - Include hidden files/folders.
 * @property {string[]|null}  [extensions=null]      - Restrict to these extensions (e.g. ['.txt']).
 * @property {number}         [minSize=0]            - Skip files below this byte count.
 * @property {number|null}    [maxSize=null]         - Skip files above this byte count.
 */

/**
 * Collect all files under `rootDir` that match the given options.
 *
 * @param {string}      rootDir - Directory to scan.
 * @param {ScanOptions} [opts]  - Filter options.
 * @returns {Promise<string[]>} Sorted array of absolute file paths.
 */
export async function collectFiles(rootDir, opts = {}) {
  const {
    recursive = false,
    includeHidden = false,
    extensions = null,
    minSize = 0,
    maxSize = null,
  } = opts;

  const absRoot = resolve(rootDir);

  const extSet = extensions
    ? new Set(
        extensions.map((e) =>
          e.startsWith(".") ? e.toLowerCase() : `.${e.toLowerCase()}`,
        ),
      )
    : null;

  const results = [];

  async function walk(dir) {
    let entries;
    try {
      entries = await readdir(dir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      const fullPath = join(dir, entry.name);

      if (entry.isSymbolicLink()) continue;

      if (entry.isDirectory()) {
        if (!recursive) continue;
        if (!includeHidden && entry.name.startsWith(".")) continue;
        if (EXCLUDED_DIRS.has(entry.name)) continue;
        await walk(fullPath);
        continue;
      }

      if (!entry.isFile()) continue;

      // Hidden files
      if (!includeHidden && entry.name.startsWith(".")) continue;

      // Excluded filenames
      if (EXCLUDED_FILES.has(entry.name)) continue;

      // Extension
      const dotIdx = entry.name.lastIndexOf(".");
      const ext = dotIdx >= 0 ? entry.name.slice(dotIdx).toLowerCase() : "";

      if (EXCLUDED_EXTS.has(ext)) continue;
      if (extSet && !extSet.has(ext)) continue;

      // Size filters
      if (minSize > 0 || maxSize !== null) {
        try {
          const info = await stat(fullPath);
          if (info.size < minSize) continue;
          if (maxSize !== null && info.size > maxSize) continue;
        } catch {
          // include anyway; detectFile will handle the error
        }
      }

      results.push(fullPath);
    }
  }

  await walk(absRoot);
  return results.sort();
}
